torchvision/models/mnasnet.py

```diff
@@ -0,0 +1,163 @@
+    assert 0.0 < round_up_bias < 1.0
+    new_val = max(divisor, int(val + divisor / 2) // divisor * divisor)
+    return new_val if new_val >= round_up_bias * val else new_val + divisor
+
+
+def _scale_depths(depths, alpha):
+    """ Scales tensor depths as in reference MobileNet code, prefers rouding up
+    rather than down. """
+    return [_round_to_multiple_of(depth * alpha, 8) for depth in depths]
+
+
+class MNASNet(torch.nn.Module):
+    """ MNASNet, as described in https://arxiv.org/pdf/1807.11626.pdf.
+    >>> model = MNASNet(1000, 1.0)
+    >>> x = torch.rand(1, 3, 224, 224)
+    >>> y = model.forward(x)
+    >>> y.dim()
+    1
+    >>> y.nelement()
+    1000
+    """
+
+    def __init__(self, num_classes, alpha, dropout=0.2):
+        super(MNASNet, self).__init__()
+        self.alpha = alpha
+        self.num_classes = num_classes
+        self.dropout = dropout
+        depths = _scale_depths([24, 40, 80, 96, 192, 320], alpha)
+        layers = [
+            # First layer: regular conv.
+            nn.Conv2d(3, 32, 3, padding=1, stride=2, bias=False),
+            nn.BatchNorm2d(32, momentum=_BN_MOMENTUM),
+            nn.ReLU(inplace=True),
+            # Depthwise separable, no skip.
+            nn.Conv2d(32, 32, 3, padding=1, stride=1, groups=32, bias=False),
+            nn.BatchNorm2d(32, momentum=_BN_MOMENTUM),
+            nn.ReLU(inplace=True),
+            nn.Conv2d(32, 16, 1, padding=0, stride=1, bias=False),
+            nn.BatchNorm2d(16, momentum=_BN_MOMENTUM),
+            # MNASNet blocks: stacks of inverted residuals.
+            _stack(16, depths[0], 3, 2, 3, 3, _BN_MOMENTUM),
+            _stack(depths[0], depths[1], 5, 2, 3, 3, _BN_MOMENTUM),
+            _stack(depths[1], depths[2], 5, 2, 6, 3, _BN_MOMENTUM),
+            _stack(depths[2], depths[3], 3, 1, 6, 2, _BN_MOMENTUM),
+            _stack(depths[3], depths[4], 5, 2, 6, 4, _BN_MOMENTUM),
+            _stack(depths[4], depths[5], 3, 1, 6, 1, _BN_MOMENTUM),
+            # Final mapping to classifier input.
+            nn.Conv2d(depths[5], 1280, 1, padding=0, stride=1, bias=False),
+            nn.BatchNorm2d(1280, momentum=_BN_MOMENTUM),
+            nn.ReLU(inplace=True),
+        ]
+        self.layers = nn.Sequential(*layers)
+        self.classifier = nn.Linear(1280, self.num_classes)
+
+        self._initialize_weights()
+
+    def features(self, x):
+        return self.layers(x)
+
+    def forward(self, x):
+        x = self.features(x)
+        # Equivalent to global avgpool and removing H and W dimensions.
+        x = x.mean([2, 3])
+        if self.dropout > 0.0:
+            x = nn.functional.dropout(x, p=self.dropout, training=self.training,
+                                      inplace=True)
+        return self.classifier(x)
+
+    def _initialize_weights(self):
+        for m in self.modules():
+            if isinstance(m, nn.Conv2d):
+                n = m.kernel_size[0] * m.kernel_size[1] * m.out_channels
```

Comment:
can you use `nn.init` methods here and in the next cases? They can be probably replaced by `kaiming_normal_` or `xavier_normal_`.

Reply:
Maybe, maybe not. I'd need to re-train with kaiming_normal_ to know if it's kosher here or not. I think I trained with either of those two (don't recall which) and couldn't get to the accuracy I achieved with this. I'll kick off a run on 0.75 with kaiming_normal_ to see if it's the initialization or something else.

Reply:
It is possible to keep the same initialization hyperparameters parameters while using `kaiming_normal_` / `xavier_normal_`, all you'd need to do is to find the corresponding `a` / `gain` that is needed to match values here.

Reply:
Actually it looks like you're right. The exact example from the documentation looks like it's identical mathematically: `nn.init.kaiming_normal_(w, mode='fan_out', nonlinearity='relu')`. It just didn't occur to me to use `mode='fan_out'`. 

I have replaced the initialization with `kaiming_normal_` and used the `nn.init` module directly, like in mobilenetv2 in this repo.